File: ios/YamabikoChat/Networking/Models/ProviderError.swift
import Foundation

enum ProviderClientError: LocalizedError, Sendable {
    case missingCredential(String)
    case invalidBaseURL(String)
    case invalidResponse
    case httpStatus(Int, String)
    case providerFailure(statusCode: Int?, code: String?, message: String)
    case parseFailure(String)
    case unsupportedModel(provider: String, model: String)

    var errorDescription: String? {
        switch self {
        case let .missingCredential(provider):
            if provider.uppercased() == "GEMINI_OAUTH_CLIENT" {
                return "API credential is missing for the selected provider."
            }
            return "Missing credential for \(provider)."
        case let .invalidBaseURL(value):
            return "Invalid base URL: \(value)."
        case .invalidResponse:
            return "Invalid API response."
        case let .httpStatus(status, body):
            return "HTTP \(status): \(body)"
        case let .providerFailure(statusCode, code, message):
            let status = statusCode.map { "HTTP \($0)" } ?? "Provider error"
            let suffix = code?.trimmedNonEmpty.map { " [\($0)]" } ?? ""
            return "\(status)\(suffix): \(message)"
        case let .parseFailure(reason):
            return "Response parse failed: \(reason)"
        case let .unsupportedModel(provider, model):
            return "Unsupported model for \(provider): \(model)."
        }
    }
}

File: app/src/main/java/com/porarri/yamabikochat/data/model/ProviderError.kt
package com.porarri.yamabikochat.data.model

sealed class ProviderClientError(message: String) : Exception(message) {
    data class MissingCredential(val provider: String) :
        ProviderClientError(
            if (provider.uppercase() == "GEMINI_OAUTH_CLIENT")
                "API credential is missing for the selected provider."
            else
                "Missing credential for $provider."
        )

    data class InvalidBaseURL(val url: String) :
        ProviderClientError("Invalid base URL: $url.")

    object InvalidResponse :
        ProviderClientError("Invalid API response.")

    data class HttpStatus(val status: Int, val body: String) :
        ProviderClientError("HTTP $status: $body")

    data class ParseFailure(val reason: String) :
        ProviderClientError("Response parse failed: $reason")

    data class UnsupportedModel(val provider: String, val model: String) :
        ProviderClientError("Unsupported model for $provider: $model.")
}
